ls.rbを追加

wak9102 opened the pull request:
ls コマンドを作るの課題提出のため、05.lsにls.rbを追加しました

wak9102 committed:
first commit

wak9102 committed:
ls.rbを追加

「ls コマンドを作る」の提出物作成のため、05.lsにls.rbを追加

wak9102 committed:
ls.rbにaオプションとrオプションの処理を追加

aオプションがなければドットファイルを削除し、rオプションがあれば降順に並び替えるため追加しました

wak9102 committed:
ls.rbにlオプションの処理を追加

lオプションがある場合、ロングフォーマットで出力する処理を追加

wak9102 committed:
ls.rbにlオプションがない場合の処理を追加

lオプションがない場合、指定の最大列数でファイル名を出力する処理を追加しました

wak9102 committed:
ls.rbでmainメソッドを追加

共通の処理をmainメソッドとして定義し、上から下に読みやすく変更

wak9102 committed:
ls.rbのaオプションの処理を変更

aオプションがない場合ではなくある場合に追加の処理をするように変更

wak9102 committed:
lオプションの出力結果を修正

ブロック数ではなくハードリンク数の表示に修正

wak9102 committed:
ls.rbに空行を追加

複数行続いている部分を読みやすくするため空行を追加

wak9102 committed:
rオプションの処理メソッドを削除

rオプションの処理をreverse!メソッドに代替し行数を削減

wak9102 committed:
aオプションのメソッドを削除

aオプションの処理が少ないためメソッドを削除し行数を削減

wak9102 committed:
ls.rbのコメントの誤字を修正

出力時の最大行数が３と記載していたので列数に修正

yoshitsugu commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,104 @@
+require 'pathname'
+require 'etc'
+
+MAX_OUTPUT_COLUMNS = 3 # lオプションがない場合の出力列数は「3」で設定
```
定数化ナイスです :+1:

yoshitsugu commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,104 @@
+  # コマンドラインオプションを配列で取得する
+  options = []
+  opts = OptionParser.new
+  opts.on('-a', '--all') { options << :all }
+  opts.on('-l') { options << :long_format }
+  opts.on('-r', '--reverse') { options << :reverse }
```
symbol使えているのよいですね :+1:

yoshitsugu commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,104 @@
+end
+
+def puts_with_long_format(files)
+  block_size = files.each.sum { |file| File.stat(file).blocks }
```
Arrayがsumを持っているので、ここ `each` 不要ですね。
https://docs.ruby-lang.org/ja/latest/method/Array/i/sum.html

wak9102 replied:
ご教示ありがとうございます🙇‍♂️
ドキュメントを確認し、eachを削除しました。

yoshitsugu reviewed (commented):
読みやすくかけているかなーと思います！ :+1:
細かいところコメントしましたので、余裕あれば確認してみてください。

yoshitsugu commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,104 @@
+
+# lオプションがない場合の出力処理
+def puts_without_long_format(files)
+  max_word_count = files.map(&:size).max # 表示するファイル名のうち最長文字数を取得
```
全てのファイルのうちの最大文字数だと、以下のように1つだけファイル名がながいときに2列目が間延びしてしまいます。余裕あれば同じ列のついの最大をとるようにしてみてください〜
![Screenshot from 2021-06-26 07-57-19](https://user-images.githubusercontent.com/748445/123715033-72fe9f80-d8b2-11eb-93fe-427b5d4caa30.png)

wak9102 replied:
ご教示ありがとうございます🙏
修正してみましたのでご確認よろしくお願いいたします！
<img width="508" alt="スクリーンショット 2021-07-02 20 20 40" src="https://user-images.githubusercontent.com/70687543/124267390-25648a00-db73-11eb-9708-d70833771c08.png">

wak9102 committed:
ls.rbのArrayのeachメソッドを削除

71行目のArray.each.sumにおいてeachは不要なため削除

wak9102 committed:
ls.rbのファイル名の出力方法を変更

lオプションがない場合の出力において、ljustの引数が「全てのファイルのうちの最大文字数」だと他の列が間延びする可能性があるので、「同列中の最大文字数」を適用するように変更

yoshitsugu commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,108 @@
+  output_rows = files.size.fdiv(MAX_OUTPUT_COLUMNS).ceil
+  # カラムごとのファイル名の最大文字数を配列で取得
+  max_word_counts = files.each_slice(output_rows).map do |files_sliced|
+    files_sliced.map(&:size).max
```
maxはブロックもとれて `files_sliced.max(&:size)` でいけると思います！
https://docs.ruby-lang.org/ja/latest/method/Array/i/max.html

wak9102 replied:
ありがとうございます🙇‍♂️
今後の参考にさせていただきます！

yoshitsugu reviewed (approved):
OKです！ :sparkles: 
1点だけ書き方でコメントしたので、参考までに。
